### PR TITLE
AT1-122: Add test cases

### PR DESCRIPTION
**Pull request by apurvapatkar**
Added test cases. The changes for coding standard compliance (story_AT1-124) are also incorporated in this branch.

### CODE REVIEW

File: tmc-prototype/DishMaster/DishMaster/DishMaster.py [hunk 1]
```diff
@@ -7,33 +7,25 @@
...
 # PyTango imports
-import PyTango
-from PyTango import DebugIt
-from PyTango.server import run
-from PyTango.server import Device, DeviceMeta
-from PyTango.server import attribute, command
-from PyTango.server import device_property
-from PyTango import AttrQuality, DispLevel, DevState
-from PyTango import AttrWriteType, PipeWriteType
+import tango
```
Comment (lochanitis) [inline tmc-prototype/DishMaster/DishMaster/DishMaster.py]: It should be _# tango imports_ now

File: tmc-prototype/DishMaster/DishMaster/DishMaster.py [hunk 2]
```diff
@@ -42,129 +34,127 @@ class DishMaster(SKAMaster):
...
-
-        for position in range(0,temp):
-            self.set_status("Dish is pointing towards the desired coordinates.")
-            self.devlogmsg("Dish is pointing towards the desired coordinates.", 4)
+            max_increment = input_increment
+        if input_increment == max_increment:
+            input_increment = input_increment + 1
+        for position in range(0, input_increment):
+            self.set_status(CONST.STR_DISH_POINT_INPROG)
+            self.devlogmsg(CONST.STR_DISH_POINT_INPROG, int(tango.LogLevel.LOG_INFO))
```
Comment (lochanitis) [inline tmc-prototype/DishMaster/DishMaster/DishMaster.py]: Make sure it is changed to _dev_logging_ everywhere.

File: tmc-prototype/SubarrayNode/SubarrayNode/SubarrayNode.py [hunk 3]
```diff
@@ -111,228 +112,174 @@ def is_EndScan_allowed(self):
...
                 self._dish_leaf_node_group.add(self.DishLeafNodePrefix +  argin[leafId])
-                devProxy = PyTango.DeviceProxy(self.DishLeafNodePrefix + argin[leafId])
+                devProxy = tango.DeviceProxy(self.DishLeafNodePrefix + argin[leafId])
                 self._dish_leaf_node_proxy.append(devProxy)
-                self._event_id = devProxy.subscribe_event("dishHealthState",
-                                                        PyTango.EventType.CHANGE_EVENT,
-                                                        self.setHealth,
-                                                        stateless=True)
+                self._event_id = devProxy.subscribe_event(CONST.EVT_DISH_HEALTH_STATE,
+                                                          tango.EventType.CHANGE_EVENT,
```
Comment (adityadangeska) [inline tmc-prototype/SubarrayNode/SubarrayNode/SubarrayNode.py]: check indentation

File: tmc-prototype/SubarrayNode/SubarrayNode/SubarrayNode.py [hunk 4]
```diff
@@ -111,228 +112,174 @@ def is_EndScan_allowed(self):
...
-                    self._read_activity_message = "Health state of " + str(evt.device) + " :-> FAILED"
-
-                elif (self._dish_health_state == 3):
-                    print "Health state of " + str(evt.device) + " :-> UNKNOWN"
-                    self._read_activity_message = "Health state of " + str(evt.device) + " :-> UNKNOWN"
-
+                if self._dish_health_state == CONST.ENUM_OK:
+                    print CONST.STR_HEALTH_STATE + str(evt.device) + CONST.STR_OK
+                    self._read_activity_message = CONST.STR_HEALTH_STATE + str(evt.device
+                                                                               ) + CONST.STR_OK
```
Comment (adityadangeska) [inline tmc-prototype/SubarrayNode/SubarrayNode/SubarrayNode.py]: Check and modify indentation

File: tmc-prototype/SubarrayNode/SubarrayNode/SubarrayNode.py [hunk 5]
```diff
@@ -111,228 +112,174 @@ def is_EndScan_allowed(self):
...
+                        ok_health_count = ok_health_count + 1
+                if ok_health_count == len(self.dishHealthStateMap.values()):
                     self._health_state = 0
-
-                elif failed != 0:
+                elif failed_health_count != 0:
                     self._health_state = 2
-
-                elif degraded != 0:
+                elif degraded_health_count   != 0:
```
Comment (adityadangeska) [inline tmc-prototype/SubarrayNode/SubarrayNode/SubarrayNode.py]: Remove extra spaces

File: tmc-prototype/DishMaster/DishMaster/DishMaster.py [hunk 6]
```diff
@@ -251,212 +243,216 @@ def init_device(self):
...
+        except Exception as except_occured:
+            print CONST.ERR_INIT_PROP_ATTR_DISH, self.ReceptorNumber
+            self.devlogmsg(CONST.ERR_INIT_PROP_ATTR_DISH, int(tango.LogLevel.LOG_ERROR))
+            print CONST.STR_ERR_MSG, except_occured
         # PROTECTED REGION END #    //  DishMaster.always_executed_hook
 
     def always_executed_hook(self):
         # PROTECTED REGION ID(DishMaster.always_executed_hook) ENABLED START #
-        pass
+        """ Internal construct of TANGO. """
```
[the remark this answers is not in the document]
Comment (apurvapatkar) [reply tmc-prototype/DishMaster/DishMaster/DishMaster.py]: pass statement is removed as per the coding guidelines.

File: tmc-prototype/CentralNode/CentralNode/CentralNode.py [hunk 7]
```diff
@@ -7,25 +7,20 @@
 # Distributed under the terms of the GPL license.
```
[the remark this answers is not in the document]
Comment (apurvapatkar) [reply tmc-prototype/CentralNode/CentralNode/CentralNode.py]: Done. The commit id is: 06d866cea5dc37e9f9c1bb7a3cb0b392e7331f1f

File: tmc-prototype/DishMaster/DishMaster/DishMaster.py [hunk 8]
```diff
@@ -42,152 +34,127 @@ class DishMaster(SKAMaster):
...
             try:
-                self.azimuthThread = threading.Thread(None, self.azimuth, 'DishMaster')
-                self.elevationThread = threading.Thread(None, self.elevation, 'DishMaster')
-                self.azimuthThread.start()
-                self.elevationThread.start()
+                self.azimuth_thread = threading.Thread(None, self.azimuth, 'DishMaster')
+                self.elevation_thread = threading.Thread(None, self.elevation, 'DishMaster')
+                self.azimuth_thread.start()
+                self.elevation_thread.start()
                 self._pointing_state = 1
```
[the remark this answers is not in the document]
Comment (apurvapatkar) [reply tmc-prototype/DishMaster/DishMaster/DishMaster.py]: Done.

File: tmc-prototype/DishMaster/DishMaster/DishMaster.py (hunk 2 above)
Comment (apurvapatkar) [reply tmc-prototype/DishMaster/DishMaster/DishMaster.py]: tmc-prototype is integrated with updated lmc base classes. This is there in the upcoming commit.

File: tmc-prototype/DishMaster/DishMaster/DishMaster.py [hunk 9]
```diff
@@ -469,180 +465,223 @@ def is_SetStandbyLPMode_allowed(self):
...
         # PROTECTED REGION ID(DishMaster.is_SetStandbyFPMode_allowed) ENABLED START #
-        return self.get_state() not in [DevState.UNKNOWN,DevState.DISABLE]
+        """ Checks if the SetStandbyFPMode is allowed in the current state of DishMaster. """
+        return self.get_state() not in [DevState.UNKNOWN, DevState.DISABLE]
         # PROTECTED REGION END #    //  DishMaster.is_SetStandbyFPMode_allowed
 
     @command(
-    dtype_in='str', 
-    doc_in="Timestamp at which command should be executed.", 
+        dtype_in='str',
```
[the remark this answers is not in the document]
Comment (apurvapatkar) [reply tmc-prototype/DishMaster/DishMaster/DishMaster.py]: Done.

File: tmc-prototype/CentralNode/CentralNode/CONST.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
+# In/Out command constants
+CMD_SET_STOW_MODE = "SetStowMode"
+CMD_SET_STANDBY_MODE = "SetStandbyLPMode"
+CMD_SET_OPERATE_MODE = "SetOperateMode"
+
+
```
[the remark this answers is not in the document]
Comment (apurvapatkar) [reply tmc-prototype/CentralNode/CentralNode/CONST.py]: Done.

File: tmc-prototype/CentralNode/CentralNode/CentralNode.py [hunk 11]
```diff
@@ -38,162 +33,85 @@ class CentralNode(SKABaseDevice):
...
-                    print "Health state of " + str(evt.device) + " :-> DEGRADED"
-                    self._read_activity_message = "Health state of " + str(evt.device) + " :-> DEGRADED"
-                elif (self._subarray_health_state == 2):
-                    print "Health state of " + str(evt.device) + " :-> FAILED"
-                    self._read_activity_message = "Health state of " + str(evt.device) + " :-> FAILED"
-                elif (self._subarray_health_state == 3):
-                    print "Health state of " + str(evt.device) + " :-> UNKNOWN"
-                    self._read_activity_message = "Health state of " + str(evt.device) + " :-> UNKNOWN"
+                if self._subarray_health_state == CONST.ENUM_OK:
+                    print CONST.STR_HEALTH_STATE + str(evt.device
```
[the remark this answers is not in the document]
Comment (apurvapatkar) [reply tmc-prototype/CentralNode/CentralNode/CentralNode.py]: The statement is aligned with coding standards.

File: tmc-prototype/CentralNode/CentralNode/CentralNode.py [hunk 12]
```diff
@@ -253,97 +161,81 @@ def push_event(self, evt):
...
-
+                subarray_proxy.subscribe_event(CONST.EVT_SUBSR_SA_HEALTH_STATE,
+                                               EventType.CHANGE_EVENT,
+                                               self.subarrayHealthStateCallback, stateless=True)
+            except Exception as except_occured:
+                print CONST.ERR_SUBSR_SA_HEALTH_STATE, self.TMMidSubarrayNodes[subarray]
+                self._read_activity_message = CONST.ERR_SUBSR_SA_HEALTH_STATE + str(self.TMMidSubarrayNodes[
+                                                                                        subarray])
+                self.devlogmsg(CONST.ERR_SUBSR_SA_HEALTH_STATE, int(tango.LogLevel.LOG_ERROR))
+                print CONST.STR_ERR_MSG, except_occured
```
[the remark this answers is not in the document]
Comment (apurvapatkar) [reply tmc-prototype/CentralNode/CentralNode/CentralNode.py]: Unnecessary print statements are already removed.

File: tmc-prototype/SubarrayNode/SubarrayNode/SubarrayNode.py (hunk 3 above)
Comment (apurvapatkar) [reply tmc-prototype/SubarrayNode/SubarrayNode/SubarrayNode.py]: Done.

File: tmc-prototype/SubarrayNode/SubarrayNode/SubarrayNode.py (hunk 4 above)
Comment (apurvapatkar) [reply tmc-prototype/SubarrayNode/SubarrayNode/SubarrayNode.py]: Done.

File: tmc-prototype/SubarrayNode/SubarrayNode/SubarrayNode.py (hunk 5 above)
Comment (apurvapatkar) [reply tmc-prototype/SubarrayNode/SubarrayNode/SubarrayNode.py]: Done.

File: tmc-prototype/CentralNode/CentralNode/CONST.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+STR_ERR_MSG = "Error message is: "
+STR_STOW_CMD_ISSUED_CN = "STOW command invoked from Central node on the requested dishes"
+STR_STANDBY_CMD_ISSUED = "StandByTelescope command invoked from Central node"
+STR_STARTUP_CMD_ISSUED = "StartUpTelescope command invoked from Central node"
+STR_INIT_SUCCESS = "CentralNode is initialized successfully."
+STR_FALSE = "False"
+
+
+#PROPERTIES DEFAULT VALUES
+PROP_DEF_VAL_TM_MID_SA1 = "tango://apurva-pc:10000/ska_mid/tm_subarray_node/1"
```
[the remark this answers is not in the document]
Comment (apurvapatkar) [reply tmc-prototype/CentralNode/CentralNode/CONST.py]: Updated. This is there in the upcoming commit.

File: tmc-prototype/CentralNode/CentralNode/CentralNode.py [hunk 14]
```diff
@@ -38,162 +33,85 @@ class CentralNode(SKABaseDevice):
...
-                elif "tm_subarray_node/2" in evt.attr_name:
+                elif CONST.PROP_DEF_VAL_TM_MID_SA2 in evt.attr_name:
                     self._subarray2_health_state = self._subarray_health_state
                 else:
-                    print "Event from the Unknown Subarray device!"
-                    self._read_activity_message = "Event from the Unknown Subarray device!"
-
+                    print CONST.EVT_UNKNOWN_SA
+                    self._read_activity_message = CONST.EVT_UNKNOWN_SA
                 self.subarrayHealthStateMap[evt.device] = self._subarray_health_state
```
[the remark this answers is not in the document]
Comment (apurvapatkar) [reply tmc-prototype/CentralNode/CentralNode/CentralNode.py]: Done.

File: tmc-prototype/CentralNode/conftest.py [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+@pytest.fixture(scope="class")
+def tango_context(request):
+    """Creates and returns a TANGO DeviceTestContext object.
+
+    Parameters
+    ----------
+    request: _pytest.fixtures.SubRequest
+        A request object gives access to the requesting test context.
+    """
+    print "request is:", request
```
[the remark this answers is not in the document]
Comment (apurvapatkar) [reply tmc-prototype/CentralNode/conftest.py]: Done for conftest.py files in all the control nodes and DishMaster. The changes are there in the upcoming commit.

File: tmc-prototype/DishMaster/test/DishMaster_test.py [hunk 16]
```diff
@@ -143,7 +128,7 @@ def test_SetMaintenanceMode(self, tango_context):
     def test_SetOperateMode(self, tango_context):
         """Test for SetOperateMode"""
         # PROTECTED REGION ID(DishMaster.test_SetOperateMode) ENABLED START #
-        #self.device.SetOperateMode()
+        tango_context.device.SetStandbyLPMode()
         tango_context.device.SetOperateMode()
         assert tango_context.device.adminMode == 0
         assert tango_context.device.dishMode == 8
```
[the remark this answers is not in the document]
Comment (apurvapatkar) [reply tmc-prototype/DishMaster/test/DishMaster_test.py]: It works without and statement.

File: tmc-prototype/DishMaster/DishMaster/DishMaster.py [hunk 17]
```diff
@@ -42,152 +34,127 @@ class DishMaster(SKAMaster):
...
         self._pointing_state = 1
         self._azimuth_difference = self._desired_pointing[1] - self._achieved_pointing[1]
-        if self._azimuth_difference > 0.0:
-            self.increment_position([1,self._azimuth_difference])
-        elif self._azimuth_difference < 0.0:
-            self.decrement_position([1,abs(self._azimuth_difference)])
-        else:
-            pass
+        if self._azimuth_difference > 0.00:
+            self.increment_position([1, self._azimuth_difference])
```
[the remark this answers is not in the document]
Comment (apurvapatkar) [reply tmc-prototype/DishMaster/DishMaster/DishMaster.py]: Done. The changes are there in the upcoming commit.

File: tmc-prototype/DishMaster/DishMaster/DishMaster.py [hunk 18]
```diff
@@ -240,8 +207,6 @@ def checkSlew(self):
         unit="km/h",
```
[the remark this answers is not in the document]
Comment (apurvapatkar) [reply tmc-prototype/DishMaster/DishMaster/DishMaster.py]: I checked this. It is the unit of WindSpeed attribute.

File: tmc-prototype/DishMaster/DishMaster/DishMaster.py [hunk 19]
```diff
@@ -276,376 +243,445 @@ def init_device(self):
...
-            self._capturing = True                      # set Capturing to True
-            self._pointing_state = 3                    # set pointingState to SCAN
-            self.set_status("Data Capturing started.")
-            self.devlogmsg("Data Capturing started", 4)
-        except Exception as e:
-            print "Unexpected error in executing StartCapture Command on Dish", self.ReceptorNumber
-            self.devlogmsg("Unexpected error in executing StartCapture Command on Dish", 2)
-            print "Error message is: \n", e
-        pass
+            if type(float(argin)) == float:
```
[the remark this answers is not in the document]
Comment (apurvapatkar) [reply tmc-prototype/DishMaster/DishMaster/DishMaster.py]: This condition raises an exception if invalid inputs are provided in the function.

File: tmc-prototype/DishMaster/DishMaster/DishMaster.py [hunk 20]
```diff
@@ -276,376 +243,445 @@ def init_device(self):
...
-            self._capturing = False                     # set Capturing to FALSE
-            self._pointing_state = 0                    # set pointingState to READY
-            self.set_status("Data Capturing stopped.")
-            self.devlogmsg("Data Capturing stopped.", 4)
-        except Exception as e:
-            print "Unexpected error in executing StopCapture Command on Dish", self.ReceptorNumber
-            self.devlogmsg("Unexpected error in executing StopCapture Command on Dish", 2)
-            print "Error message is: \n", e
-        pass
+            if type(float(argin)) == float:
```
[the remark this answers is not in the document]
Comment (apurvapatkar) [reply tmc-prototype/DishMaster/DishMaster/DishMaster.py]: This condition raises an exception if invalid inputs are provided in the function.

File: tmc-prototype/DishMaster/DishMaster/DishMaster.py [hunk 21]
```diff
@@ -276,376 +243,445 @@ def init_device(self):
...
+        input timestamp.
+        :param argin: timestamp
+        :return: None
+        """
         try:
             # Command to start SCAN
-            if (self._pointing_state == 0):
+            if self._pointing_state == 0:
                 self._current_time = time.time()
                 self._scan_execution_time = float(argin)
```
[the remark this answers is not in the document]
Comment (jyotinpo) [reply tmc-prototype/DishMaster/DishMaster/DishMaster.py]: _argin_ here represent the timestamp at which the scan shall be executed. We have checked and found that the timestamp can also be a floating point number. Hence, keeping it the same.

File: tmc-prototype/DishMaster/DishMaster/DishMaster.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (jyotinpo) [reply tmc-prototype/DishMaster/DishMaster/DishMaster.py]: We tried various options and found that the logic used in the current code is perfectly fine. Hence, keeping the _if_ condition same. Please check below practice code:

`>>> str = "a"
>>> if float(str)!= None:
...  print "True"
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: could not convert string to float: a
>>> str = 2
>>> if float(str)!= None:
...  print "True"
... 
True
>>> float('nan')
nan
>>> float('n')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: could not convert string to float: n
>>> str = "a"
>>> if float(str) != nan:
...  print "OK"
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: could not convert string to float: a
>>> str = 5
>>> if float(str) != nan:
...  print "OK"
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'nan' is not defined
>>> if float(str) != float('nan'):
...  print "OK"
... 
OK
>>> str = "a"
>>> if float(str) != float('nan'):
...  print "OK"
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: could not convert string to float: a
>>> str = 5
>>> if float(str) != float('nan'):
...  print "OK"
... 
OK
>>> str = "nan"
>>> if float(str) != float('nan'):
...  print "OK"
... 
OK
>>> 
`

File: tmc-prototype/DishMaster/DishMaster/DishMaster.py [hunk 22]
```diff
@@ -42,152 +34,127 @@ class DishMaster(SKAMaster):
...
-            self.set_status("Dish is pointing towards the desired coordinates.")
-            self.devlogmsg("Dish is pointing towards the desired coordinates.", 4)
+            max_increment = input_increment
+        if input_increment == max_increment:
+            input_increment = input_increment + 1
+        for position in range(0, input_increment):
+            self.set_status(CONST.STR_DISH_POINT_INPROG)
+            self.devlogmsg(CONST.STR_DISH_POINT_INPROG, int(tango.LogLevel.LOG_INFO))
             self._pointing_state = 1
             time.sleep(2)
```
[the remark this answers is not in the document]
Comment (jyotinpo) [reply tmc-prototype/DishMaster/DishMaster/DishMaster.py]: Sleep time was kept 2 seconds to demonstrate pointing clearly. Keeping it same.

**Pull request merged**